[WIP] refactor sync/analyze stage 1a

thearthur opened the pull request:
first pass, this is incomplete because the new analysis is not plumbed
into all the places where we actually need to run analyze.

salsakran commented on src/metabase/cache_database.clj:
```diff
@@ -0,0 +1,76 @@
+(ns metabase.cache-database
```
this should be either "cache-values" or "field-values" or "cached-field-values". any opinions @camsaul ?

salsakran replied:
or maybe "cache-field-values"

camsaul replied:
yeah. it shouldn't be a top-level namespace either IMO. It should be `metabase.sync.something`. That way it's clear it's part of the sync part of the app.

@thearthur think about that chart you drew up showing how the different parts of the app interconnect, and how much more confusing that would have been to make if everything was just a top-level namespace

salsakran reviewed (commented):
couple things jump out at me

1. the naming of the "cache" section is confusing. We're *not* caching databases, tables or fields here. We're just extracting a set of field values for a subset of fields in each of the above. We should be explicit about this.
2. It seems like this re-fetches field values for both analyze and field-value-caching. we should figure out a way to make it possible to chain these together on the first run of things. 
eg. the first time I run analysis on a freshly added db 
I'm going to do both a sync (that doesn't need to touch field values), and then a fingerprint and a field-values-cache (which do). The latter two should be chainable off a single fetch of field values.

thearthur committed:
split analize into analize and classify

(second attempt)

thearthur committed:
Improve the error message when using wrong ssh tunnel host

Or when running the tests on a computer thats not running sshd

thearthur committed:
further split sync and cache-fields

thearthur committed:
descriptive-fingerprints-part1

thearthur committed:
use only fingerprints to classify fields

thearthur committed:
move speceial type inference to classify

thearthur committed:
working on tests for sync/analyze refactor

salsakran commented on test/metabase/sync_database/analyze_test.clj:
```diff
@@ -18,18 +18,6 @@
 ;; test:cardinality-and-extract-field-values
 ;; (#2332) check that if field values are long we skip over them
-(expect
-  {:values nil}
```
why are these deleted?

thearthur replied:
moved to metabase.sync-database.cached-values-test

salsakran commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -157,86 +53,42 @@
+       :field-percent-email     (if (values-are-valid-emails? values) 100 0)
+       :field-avg-length        (u/try-apply (:field-avg-length driver) field)
+       :visibility_type         (:visibility_type field)
+       :values                  values
```
Values should not be in the fingerprint. They will be both large and potentially sensitive.

thearthur replied:
left that in while debugging the sync tests  (which are still broken)

thearthur committed:
fix bug in fieldvalue scheduled caching

code cleanup and comments as well

thearthur committed:
further test fixing

thearthur committed:
continuing cleanup of sample values for email,json,url percent

salsakran commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -157,86 +53,53 @@
+                     (driver/field-values-lazy-seq driver field))]
+    {:id                      (:id field) ;; check if this should be field or table id
+     :field-percent-urls      (percent-valid-urls values)
+     :field-percent-json      (if (values-are-valid-json? values) 100 0)
```
this should return the percent, not turn it into a binary feature.

thearthur committed:
supply default field-avg-length function in driver

thearthur commented:
The new approach to calculating the field "fingerprint" is to remove all the driver specific logic and derive all the field fingerprint values from the first 1000 entries for the field in the table.

thearthur committed:
make notify trigger re-analysis

We should decide if we want this to happen, and for now it
fixes some tests that make this assumption.

thearthur committed:
remove pecent url calculation from driver

This was a source of full table scans.
More progress towards restoring broken tests

salsakran commented on src/metabase/driver.clj:
```diff
@@ -20,7 +20,7 @@
    This many is probably fine for inferring special types and what-not; we don't want
    to scan millions of values at any rate."
-  10000)
+  1000)
```
we should experiment with this.

thearthur committed:
split guessing special type by name to it's own namespace

This method is used on both in several places, and it's really independent.

thearthur committed:
wip

thearthur committed:
still conslidating places where decissions to cache values are made

thearthur committed:
include primary and foreign keys in fingerprint

thearthur committed:
wip cleanup

thearthur committed:
Fix tests where test data must have cached values

thearthur commented:
This is not the most current PR for this issue. new conversation should move to: https://github.com/metabase/metabase/pull/5363
